### DISCUSSION TITLE
Question about applying OSM file relationship for Valhalla engine

### DISCUSSION THREAD

**Discussion by ChanWhanPark** [valhalla/valhalla#4058] [Q&A] [answered]
Hello. We are developing another project through Valhalla Engine.
In the process of applying the OSM file, ask if the no_entry relationship is recognized in the following cases.

![image](https://user-images.githubusercontent.com/68843687/229699547-05a6706e-b3a9-4065-b16c-505d32471760.png)

In the following image, the OSM relationship was defined as follows in order to allow a left turn and not a U-turn.
![image](https://user-images.githubusercontent.com/68843687/229700143-91cccc2b-9acc-46b7-a41e-527bfd115ce2.png)

However, the valhalla engine does not recognize the OSM relationship with multiple vias, so both left and U-turns are currently possible in that example.

I'm asking if I applied OSM data correctly or if the current engine is not aware of multi-vias.

Thanks.

**Comment by kevinkreiser** [comment, marked as answer]
we call those complex restrictions and yes we do support them. looking at your tagging i think you may have made a mistake. i dont think you need to tag the nodes as members of the restriction but rather just the ways. perhaps it is this tagging that confuses valhalla. see here: https://wiki.openstreetmap.org/wiki/Relation:restriction#Members

> As shown to the left, a turn restriction can either have one node with a via role, or one or more ￼ way(s) as members with a ￼ via role.

i read this as you can tag its via members with EITHER 1 node OR 1 or more ways BUT NOT both
